segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainerMetadata.java

```diff
@@ -289,15 +293,26 @@ public void setContainerEpoch(long value) {
     @Override
     public Collection<SegmentMetadata> cleanup(Collection<SegmentMetadata> evictionCandidates, long sequenceNumberCutoff) {
         long adjustedCutoff = Math.min(sequenceNumberCutoff, this.lastTruncatedSequenceNumber.get());
+        HashMap<Long, Integer> txnCounts = new HashMap<>();
         Collection<SegmentMetadata> evictedSegments = new ArrayList<>(evictionCandidates.size());
         synchronized (this.lock) {
-            // Find those segments that have active transactions associated with them.
-            Set<Long> activeTransactions = getSegmentsWithActiveTransactions(adjustedCutoff);
+            // Count how many total transactions per parent segment there are.
+            this.metadataById
```

Comment:
The logic seems right to me as we count the number of active txns per parent, and next subtract the txn segments that are eviction candidates. We only really evict in the case the parent has no remaining active txn segment (active txn segment not being evicted).

The one thing I'm wondering is whether we can do better than a linear scan over `metadataById` to determine the txn counts. Can't we keep a counter per parent segment and simply look at the counter of non-txn eviction candidates?